// @keystonehq/hw-transport-usb/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keystoneUSBProductId = exports.keystoneUSBVendorId = exports.OFFSET_CDATA = exports.OFFSET_LC = exports.OFFSET_P2 = exports.OFFSET_P1 = exports.OFFSET_INS = exports.OFFSET_CLA = exports.USBTimeout = exports.MAXUSBPackets = exports.USBPackageSize = exports.USBInterfaceNumber = exports.USBConfigurationValue = void 0;
exports.USBConfigurationValue = 1;
// use 0 for production firmware
// use 1 for development firmware since the 0 interface number will be used by MSC for development
exports.USBInterfaceNumber = process.env.KEYSTONE_USB_ENV === 'development' ? 1 : 0;
exports.USBPackageSize = 64;
exports.MAXUSBPackets = 200;
// 15s to set the usb timeout since it may need users' action on the device.
exports.USBTimeout = 1000 * 15;
exports.OFFSET_CLA = 0;
exports.OFFSET_INS = 1;
exports.OFFSET_P1 = 3;
exports.OFFSET_P2 = 5;
exports.OFFSET_LC = 7;
exports.OFFSET_CDATA = 9;
// keystone 3 pro coldwallet device identifier
exports.keystoneUSBVendorId = 4617;
exports.keystoneUSBProductId = 12289;
