// @keystonehq/hw-transport-usb/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isString = exports.isUint8Array = exports.generateRequestID = exports.isEmpty = exports.setUint16 = exports.safeJSONparse = exports.safeJSONStringify = void 0;
const safeJSONStringify = (obj) => {
    try {
        return JSON.stringify(obj);
    }
    catch (err) {
        return '';
    }
};
exports.safeJSONStringify = safeJSONStringify;
const safeJSONparse = (str) => {
    try {
        return JSON.parse(str);
    }
    catch (err) {
        return str;
    }
};
exports.safeJSONparse = safeJSONparse;
const setUint16 = (array, offset, value) => {
    array[offset] = (value >> 8) & 0xFF;
    array[offset + 1] = value & 0xFF;
};
exports.setUint16 = setUint16;
const isEmpty = (value) => {
    if (Array.isArray(value)) {
        return value.length === 0;
    }
    if (typeof value === 'object') {
        return Object.keys(value).length === 0;
    }
    return !value;
};
exports.isEmpty = isEmpty;
const generateRequestID = () => {
    return Math.floor(Math.random() * 65535);
};
exports.generateRequestID = generateRequestID;
const isUint8Array = (data) => data instanceof Uint8Array;
exports.isUint8Array = isUint8Array;
const isString = (data) => typeof data === 'string';
exports.isString = isString;
