// @keystonehq/hw-app-eth/lib/new.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const uuid = __importStar(require("uuid"));
const bc_ur_registry_1 = require("@keystonehq/bc-ur-registry");
const bc_ur_1 = require("@ngraveio/bc-ur");
const hw_transport_webusb_1 = require("@keystonehq/hw-transport-webusb");
const hw_transport_error_1 = require("@keystonehq/hw-transport-error");
const bc_ur_registry_eth_1 = require("@keystonehq/bc-ur-registry-eth");
const util_1 = require("@ethereumjs/util");
const util_2 = require("./util");
const pathToKeypath = (path) => {
    const paths = path.replace(/[m|M]\//, '').split('/');
    const pathComponents = paths.map(path => {
        const index = parseInt(path.replace("'", ''), 10);
        const isHardened = path.endsWith("'");
        return new bc_ur_registry_1.PathComponent({ index, hardened: isHardened });
    });
    return new bc_ur_registry_1.CryptoKeypath(pathComponents);
};
class Eth {
    /**
     * Constructs a new instance of the class.
     *
     * @param transport - An object of type TransportWebUSB
     * @param mfp - Optional parameter of type string, default is undefined, but the mfp should exist in the signing process.
     */
    constructor(transport, mfp) {
        // Initialize Solana connection
        this.transport = transport;
        if (mfp) {
            this.mfp = mfp;
        }
    }
    precheck() {
        if (!this.transport) {
            (0, hw_transport_error_1.throwTransportError)(hw_transport_error_1.Status.ERR_TRANSPORT_HAS_NOT_BEEN_SET);
        }
        if (!this.mfp) {
            throw new Error("missing mfp for this wallet");
        }
    }
    sendToDevice(actions, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.transport.send(actions, data);
        });
    }
    checkDeviceLockStatus() {
        return __awaiter(this, void 0, void 0, function* () {
            let result = yield this.sendToDevice(hw_transport_webusb_1.Actions.CMD_CHECK_LOCK_STATUS, '');
            return result.payload;
        });
    }
    /**
     * Retrieves the Ethereum address and related information based on the specified path
     *
     * @param path - The derivation path for the address
     * @param boolDisplay - A boolean indicating whether to display the QR code (default: false)
     * @param boolChainCode - A boolean indicating whether to include the chain code (default: undefined)
     * @param chainId - The ID of the Ethereum chain (default: undefined)
     * @returns A Promise object containing the address, public key, master fingerprint, and optionally the chain code
     */
    getAddress(path, boolDisplay = false, boolChainCode, chainId) {
        return __awaiter(this, void 0, void 0, function* () {
            // Send a request to the device to get the address at the specified path
            const curve = bc_ur_registry_1.Curve.secp256k1;
            const algo = bc_ur_registry_1.DerivationAlgorithm.slip10;
            const kds = new bc_ur_registry_1.KeyDerivationSchema(pathToKeypath(path), curve, algo, "ETH");
            const keyDerivation = new bc_ur_registry_1.KeyDerivation([kds]);
            const hardwareCall = new bc_ur_registry_1.QRHardwareCall(bc_ur_registry_1.QRHardwareCallType.KeyDerivation, keyDerivation, "Keystone USB SDK", bc_ur_registry_1.QRHardwareCallVersion.V1);
            let ur = hardwareCall.toUR();
            const encodedUR = new bc_ur_1.UREncoder(ur, Infinity).nextPart().toUpperCase();
            const response = yield this.sendToDevice(hw_transport_webusb_1.Actions.CMD_RESOLVE_UR, encodedUR);
            let resultUR = parseResponoseUR(response.payload);
            let account = bc_ur_registry_1.CryptoMultiAccounts.fromCBOR(resultUR.cbor);
            let key = account.getKeys()[0];
            // reset the mfp when getting the address.
            this.mfp = account.getMasterFingerprint().toString('hex');
            // the compressed public key from 
            const pubkey = key.getKey();
            const chainCode = key.getChainCode();
            const uncompressedPubkey = (0, util_2.convertCompresskey)(pubkey.toString('hex'));
            const keyBuffer = Buffer.from(uncompressedPubkey, 'hex').slice(1);
            const ethAddress = util_1.Address.fromPublicKey(keyBuffer).toString();
            return {
                address: ethAddress,
                publicKey: uncompressedPubkey,
                chainCode: boolChainCode ? chainCode.toString('hex') : undefined,
                mfp: this.mfp,
            };
        });
    }
    sign(path, data, type) {
        return __awaiter(this, void 0, void 0, function* () {
            this.precheck();
            const encodedUR = constructURRequest(data, path, this.mfp, type);
            const response = yield this.sendToDevice(hw_transport_webusb_1.Actions.CMD_RESOLVE_UR, encodedUR);
            let resultUR = parseResponoseUR(response.payload);
            return parseSignatureUR(resultUR);
        });
    }
    /**
     * Sign a transaction using a specific path and raw transaction data
     *
     * @param path - The derivation path for the private key
     * @param rawTxHex - The raw transaction data in hexadecimal format
     * @param isLegacy - A boolean indicating whether the transaction is of the legacy type (default: false)
     * @returns A Promise object containing the signature data (r, s, v)
     * @throws Error if unable to sign the transaction
     */
    signTransaction(path, rawTxHex, isLegacy = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const txBuffer = Buffer.from(rawTxHex, 'hex');
            if (isLegacy) {
                return this.sign(path, txBuffer, bc_ur_registry_eth_1.DataType.transaction);
            }
            return this.sign(path, txBuffer, bc_ur_registry_eth_1.DataType.typedTransaction);
        });
    }
    /**
     * Signs an EIP-712 message using a specific path and the JSON representation of the message.
     *
     * This function takes a derivation path and a JSON object representing the message, and returns
     * a Promise that resolves to an object containing the signature values (r, s, and v).
     *
     * @param path - The derivation path for the private key used to sign the message.
     * @param jsonMessage - A JavaScript object that represents the message to be signed.
     * This object should be structured according to the format required by EIP-712.
     * @return A Promise that resolves to an object with properties:
     *         - r: A string representing the "r" value of the signature.
     *         - s: A string representing the "s" value of the signature.
     *         - v: A string representing the "v" value of the signature.
     *
    */
    signEIP712Message(path, jsonMessage) {
        return __awaiter(this, void 0, void 0, function* () {
            const messageBuffer = Buffer.from(JSON.stringify(jsonMessage), 'utf-8');
            return this.sign(path, messageBuffer, bc_ur_registry_eth_1.DataType.typedData);
        });
    }
    /**
     * Sign a personal message using a specific path and message content in hexadecimal format.
     *
     * This function takes a derivation path and a message represented by a hexadecimal string,
     * and returns a Promise that resolves to an object containing the signature values (r, s, v).
     *
     * @param path - The derivation path for the private key used to sign the message.
     * @param messageHex - A hexadecimal string that represents the message to be signed.
     * @return A Promise that resolves to an object with properties:
     *         - r: A string representing the "r" value of the signature.
     *         - s: A string representing the "s" value of the signature.
     *         - v: A string representing the "v" value of the signature.
     */
    signPersonalMessage(path, messageHex) {
        return __awaiter(this, void 0, void 0, function* () {
            const messageBuffer = Buffer.from(messageHex, "utf-8");
            return this.sign(path, messageBuffer, bc_ur_registry_eth_1.DataType.personalMessage);
        });
    }
    /**
    * Retrieves the configuration information of the connected hardware device.
    *
    * This method sends a request to the device to get its version information,
    * then parses the response to extract the firmware version and wallet master fingerprint.
    *
    * @returns A Promise that resolves to an object containing:
    *          - version: A string representing the firmware version of the device
    *          - mfp: A string representing the master fingerprint of the wallet
    * @throws Will throw an error if the device communication fails or if the response cannot be parsed
    */
    getAppConfig() {
        return __awaiter(this, void 0, void 0, function* () {
            let response = yield this.sendToDevice(hw_transport_webusb_1.Actions.CMD_GET_DEVICE_VERSION, '');
            let result = response.payload;
            let appConfig = JSON.parse(result);
            return {
                version: appConfig['firmwareVersion'],
                mfp: appConfig['walletMFP']
            };
        });
    }
}
exports.default = Eth;
const constructURRequest = (txBuffer, path, mfp, type) => {
    const requestId = uuid.v4();
    let urRequest = bc_ur_registry_eth_1.EthSignRequest.constructETHRequest(txBuffer, type, path, mfp, requestId);
    const ur = urRequest.toUR();
    const encodedUR = new bc_ur_1.UREncoder(ur, Infinity).nextPart().toUpperCase();
    return encodedUR;
};
const parseResponoseUR = (urPlayload) => {
    const decoder = new bc_ur_1.URDecoder();
    decoder.receivePart(urPlayload);
    if (!decoder.isComplete()) {
        (0, hw_transport_error_1.throwTransportError)(hw_transport_error_1.Status.ERR_UR_INCOMPLETE);
    }
    const resultUR = decoder.resultUR();
    return resultUR;
};
const parseSignatureUR = (ur) => {
    let signature = bc_ur_registry_eth_1.ETHSignature.fromCBOR(ur.cbor);
    let signatureBuffer = signature.getSignature();
    const r = signatureBuffer.slice(0, 32).toString('hex');
    const s = signatureBuffer.slice(32, 64).toString('hex');
    ;
    const v = signatureBuffer.slice(64).toString('hex');
    ;
    return {
        r,
        s,
        v
    };
};
